### DISCUSSION TITLE
How can I run the startup() method only one time ?

### DISCUSSION THREAD

**Discussion by ChrisSAV-scal** [pymeasure/pymeasure#448] [Q&A]
Hi, 
I'd like to run only one time the startup() method in the experiment execution  and afterwards loop  execute() method several times 
What is the best way to perform that ?

Thanks in advance
CS

**Comment by neal-kepler** [comment]
Without knowing the exact reason you want to do this I can't give a thorough answer but I can give you the simplest solution assuming the most forgiving situation. 

If you want to do this because the startup method takes forever and is unnecessary in subsequent loops, but you can afford making a new connection to the instrument in each loop AND you want the subsequent executions to occurs as separate procedures (i.e. generate a new file for each subsequent execution) then here is a possible solution.

Use a custom gui like in this example: https://github.com/pymeasure/pymeasure/blob/master/examples/Basic/gui_custom_inputs.py.

define a boolean parameter in the parameters list `do_startup = BooleanParameter('Do startup routine')`. If you want to be able to set this from the front panel of the custom GUI, a checkbox is the right way to go and then you would access the value as `self.inputs.checkbox_name.isChecked()`. If you don't care about being able to manually set this value, you don't need to add it to the front panel.

gate your startup function with this parameter:

```python
def startup(self):
  #initialize instruments, you still need create references to the instruments, although you can do that in execution if you want
  if do_startup:
     # do startup things
```

then in the queue function:

```python 
def queue(self):        
   filename = tempfile.mktemp()
   procedure = TestProcedure()
   n = self.inputs.do_n.value() #assuming you have a spin box on your custom front panel called do_n
   procs = []
   for i in range(n):
      if i ==0:  # do the startup only on the first loop
          procedure.do_startup = True
      else:
          procedure.do_startup = False
      procedure.seed = str(self.inputs.seed.text())        
      procedure.iterations = self.inputs.iterations.value()        
      procedure.delay = self.inputs.delay.value()
      results = Results(procedure, filename)
      procs.append(results)
   for proc in procs: # queue all the experiments with only the first having the do_startup flag true
      experiment = self.new_experiment(res)
      self.manager.queue(experiment)
```

If your problem is that you can't keep creating instances of instruments (`self.instrument = XtremeMeasurer3000(12)`) then the situation is a bit more complex, you will need to create an instrument server or something that you can make and break connections with that keeps the main connection to the instrument alive.

Edit: For posterity, there is no reason you need to use a GUI either. If you are defining your procedures to be run from the command line, then of course you can set the parameters to whatever you want without having to pull them from a GUI as described [here](https://pymeasure.readthedocs.io/en/latest/tutorial/procedure.html#using-procedures).

**Comment by neal-kepler** [comment]
If you don't want the separate executions to different files, then you could just make a ` do_n =IntegerParameter` and run the code inside the execution loop just keep emitting the data. Then you won't need a custom GUI either.

**Comment by ChrisSAV-scal** [comment]
Thanks a lot for yours answers. 

In fact my problem is that I use a "complex" FPGA SW for programming and for acquisition/generation data ,  with multithreads. 
for legacy code reasons I don't want to change this SW. 
And you're right , I can't  create more one instance of this environnement
As you suggest,  I think the best way will be to create an instrument server or whatever. Can you show me the way to do that ?
Else the workaround with a boolean parameter should be enough for me.

Thanks for all
CS

**Reply by neal-kepler** [reply]
I feel this in my soul. I wrote such a server for a serial instrument that had a similar problem a long time ago based on the server interface in this repository: https://github.com/guenp/PyQDInstrument 

That repository is solving a slightly different problem (using TCP/IP to communicate across two different versions of python [IronPython vs vanilla python]) but the idea is applicable. I'll post my more purpose built thing tomorrow when I find it.

**Reply by physicsninja** [reply]
[InstrumentServer.zip](https://github.com/pymeasure/pymeasure/files/6525573/InstrumentServer.zip)
Ok, it has been a while since I looked at this but I tried to add a few helpful comments. There are three main parts:
1. The class that wraps arounds the device and exposes attributes
2. The remote version of 1) that has those same attributes and forwards them to the server
3. The server that creates and instance 1) and executes the commands it receives from 2) on the instance of 1).

Also you should be able to install it with `pip install -e .` inside the package directory. That way you can import it as serialserverKavli or whatever you rename it to in the setup.py file.

Just want to reiterate this is not a quality solution. It will likely work, but I wrote it when I was much worse at python and I am providing it as a starting point. Let me know if you run into trouble.

**Comment by ChrisSAV-scal** [comment]
Hi, 
Thanks a lot, I'll try it next week.
